File: src/lib/crdt/dotscontext.test.ts
import { describe, expect, it } from "vitest";
import DotsContext from "./dotscontext";

describe("DotsContext", () => {
    it("should be able to get next dot", () => {
        const dots = new DotsContext();

        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.next("a")).toEqual(["a", 3]);
    });

    it("should be able to get next dot for multiple ids", () => {
        const dots = new DotsContext();

        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.next("b")).toEqual(["b", 1]);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.next("b")).toEqual(["b", 2]);
        expect(dots.next("a")).toEqual(["a", 3]);
        expect(dots.next("b")).toEqual(["b", 3]);
    });

    it("should be able to get max dot", () => {
        const dots = new DotsContext();

        expect(dots.max("a")).toEqual(0);
        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.max("a")).toEqual(1);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.max("a")).toEqual(2);
        expect(dots.next("a")).toEqual(["a", 3]);
        expect(dots.max("a")).toEqual(3);
    });

    it("should be able to get max dot for multiple ids", () => {
        const dots = new DotsContext();

        expect(dots.max("a")).toEqual(0);
        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.max("a")).toEqual(1);
        expect(dots.next("b")).toEqual(["b", 1]);
        expect(dots.max("b")).toEqual(1);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.max("a")).toEqual(2);
        expect(dots.next("b")).toEqual(["b", 2]);
        expect(dots.max("b")).toEqual(2);
        expect(dots.next("a")).toEqual(["a", 3]);
        expect(dots.max("a")).toEqual(3);
        expect(dots.next("b")).toEqual(["b", 3]);
        expect(dots.max("b")).toEqual(3);
    });

    it("should be able to check if dot is in context", () => {
        const dots = new DotsContext();

        expect(dots.has("a", 1)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.has("a", 1)).toBe(true);
        expect(dots.has("a", 2)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.has("a", 2)).toBe(true);
        expect(dots.has("a", 3)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 3]);
        expect(dots.has("a", 3)).toBe(true);
    });

    it("should be able to check if dot is in context for multiple ids", () => {
        const dots = new DotsContext();

        expect(dots.has("a", 1)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 1]);
        expect(dots.has("a", 1)).toBe(true);
        expect(dots.has("b", 1)).toBe(false);
        expect(dots.next("b")).toEqual(["b", 1]);
        expect(dots.has("b", 1)).toBe(true);
        expect(dots.has("a", 2)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 2]);
        expect(dots.has("a", 2)).toBe(true);
        expect(dots.has("b", 2)).toBe(false);
        expect(dots.next("b")).toEqual(["b", 2]);
        expect(dots.has("b", 2)).toBe(true);
        expect(dots.has("a", 3)).toBe(false);
        expect(dots.next("a")).toEqual(["a", 3]);
        expect(dots.has("a", 3)).toBe(true);
        expect(dots.has("b", 3)).toBe(false);
        expect(dots.next("b")).toEqual(["b", 3]);
        expect(dots.has("b", 3)).toBe(true);
    });

    it("should be able to merge with itself", () => {
        const dots = new DotsContext();

        dots.next("a");
        dots.next("a");
        dots.next("a");

        expect(dots.merge(dots)).toEqual(dots);
    });

    it("should be able to merge with another context", () => {
        const dots1 = new DotsContext();
        const dots2 = new DotsContext();

        dots1.next("a");
        dots1.next("a");
        dots2.next("a");
        dots2.next("a");
        dots2.next("a");

        expect(dots1.merge(dots2)).toEqual(new DotsContext({ a: 3 }));
    });

    it("should be able to merge with another context for multiple ids", () => {
        const dots1 = new DotsContext();
        const dots2 = new DotsContext();

        dots1.next("a");
        dots1.next("a");
        dots2.next("b");
        dots2.next("b");
        dots2.next("b");

        expect(dots1.merge(dots2)).toEqual(new DotsContext({ a: 2, b: 3 }));
    });
});

File: server/src/test/kotlin/pt/up/fe/sdle/crdt/DotsContextTest.kt
package pt.up.fe.sdle.crdt

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

private const val ID1 = "a"
private const val ID2 = "b"

class DotsContextTest {
    @Test
    fun `DotsContext - get next dot`() {
        val dots = DotsContext()

        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
    }

    @Test
    fun `DotsContext - get next dot for multiple ids`() {
        val dots = DotsContext()

        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertEquals(Dot(ID2, 1), dots.next(ID2))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertEquals(Dot(ID2, 2), dots.next(ID2))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
        assertEquals(Dot(ID2, 3), dots.next(ID2))
    }

    @Test
    fun `DotsContext - get max dot`() {
        val dots = DotsContext()

        assertEquals(0, dots.max(ID1))
        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertEquals(1, dots.max(ID1))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertEquals(2, dots.max(ID1))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
        assertEquals(3, dots.max(ID1))
    }

    @Test
    fun `DotsContext - get max dot for multiple ids`() {
        val dots = DotsContext()

        assertEquals(0, dots.max(ID1))
        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertEquals(1, dots.max(ID1))
        assertEquals(Dot(ID2, 1), dots.next(ID2))
        assertEquals(1, dots.max(ID2))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertEquals(2, dots.max(ID1))
        assertEquals(Dot(ID2, 2), dots.next(ID2))
        assertEquals(2, dots.max(ID2))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
        assertEquals(3, dots.max(ID1))
        assertEquals(Dot(ID2, 3), dots.next(ID2))
        assertEquals(3, dots.max(ID2))
    }

    @Test
    fun `DotsContext - has dot`() {
        val dots = DotsContext()

        assertFalse(dots.has(ID1, 1))
        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertTrue(dots.has(ID1, 1))
        assertFalse(dots.has(ID1, 2))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertTrue(dots.has(ID1, 2))
        assertFalse(dots.has(ID1, 3))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
        assertTrue(dots.has(ID1, 3))
    }

    @Test
    fun `DotsContext - has dot for multiple ids`() {
        val dots = DotsContext()

        assertFalse(dots.has(ID1, 1))
        assertEquals(Dot(ID1, 1), dots.next(ID1))
        assertTrue(dots.has(ID1, 1))
        assertFalse(dots.has(ID2, 1))
        assertEquals(Dot(ID2, 1), dots.next(ID2))
        assertTrue(dots.has(ID2, 1))
        assertFalse(dots.has(ID1, 2))
        assertEquals(Dot(ID1, 2), dots.next(ID1))
        assertTrue(dots.has(ID1, 2))
        assertFalse(dots.has(ID2, 2))
        assertEquals(Dot(ID2, 2), dots.next(ID2))
        assertTrue(dots.has(ID2, 2))
        assertFalse(dots.has(ID1, 3))
        assertEquals(Dot(ID1, 3), dots.next(ID1))
        assertTrue(dots.has(ID1, 3))
        assertFalse(dots.has(ID2, 3))
        assertEquals(Dot(ID2, 3), dots.next(ID2))
        assertTrue(dots.has(ID2, 3))
    }

    @Test
    fun `DotsContext - merge with itself`() {
        val dots = DotsContext()

        dots.next(ID1)
        dots.next(ID1)
        dots.next(ID1)

        assertEquals(dots, dots.merge(dots))
    }

    @Test
    fun `DotsContext - merge with another context`() {
        val dots1 = DotsContext()
        val dots2 = DotsContext()

        dots1.next(ID1)
        dots1.next(ID1)
        dots2.next(ID1)
        dots2.next(ID1)
        dots2.next(ID1)

        assertEquals(DotsContext(mutableMapOf(ID1 to 3)), dots1.merge(dots2))
    }

    @Test
    fun `DotsContext - merge with another context for multiple ids`() {
        val dots1 = DotsContext()
        val dots2 = DotsContext()

        dots1.next(ID1)
        dots1.next(ID1)
        dots2.next(ID2)
        dots2.next(ID2)
        dots2.next(ID2)

        assertEquals(DotsContext(mutableMapOf(ID1 to 2, ID2 to 3)), dots1.merge(dots2))
    }
}
